marimo/_runtime/runtime.py

```diff
@@ -1116,6 +1117,31 @@ async def run(
             self.mutate_graph(filtered_requests, deletion_requests=[])
         )
 
+    async def rename_file(self, filename: str) -> None:
+        self.globals["__file__"] = filename
+        roots = set()
+        for cell in self.graph.cells.values():
+            if "__file__" in cell.refs:
+                roots.add(cell.cell_id)
+
+        runner = cell_runner.Runner(
+            roots=roots,
+            graph=self.graph,
+            glbls=self.globals,
+            excluded_cells=set(),
+            debugger=self.debugger,
+            execution_mode=self.reactive_execution_mode,
+            execution_type=self.execution_type,
+            execution_context=self._install_execution_context,
+            preparation_hooks=self._preparation_hooks,
+            pre_execution_hooks=self._pre_execution_hooks,
+            post_execution_hooks=self._post_execution_hooks,
+            on_finish_hooks=(
+                self._on_finish_hooks + [self._broadcast_missing_packages]
+            ),
+        )
+        await runner.run_all()
```

Comment:
I missed this in my first review. This should be:

```
await self._run_cells(roots)
```

which wraps the `Runner` and does some additional housekeeping, including handling `mo.state` updates.

Reply:
Much cleaner